### PR TITLE
Introduce an util to retrieve the ID of the first available user who can install llms stuff

### LINKED ISSUE #1577: 5.0.0 alpha issues/feedback
Feedback / Issues from @eri-trabiccolo that we haven't investigate too thoroughly yet but we're recording here to avoid potential duplicates if any other tests stumble on these issues:

+ [x] 1) radio buttons in the editor look like checkboxes - so squared and with a small border radius, it’s because of our CSS that applies a small border radius to all the fields
+ [x] 2) checkboxes: it’s not possible to remove a default option once set (in the right panel)
+ [x] 3) checkboxes in the right panel look like radio buttons, and they shouldn’t, also you can check only one option by default, and this doesn’t make a lot of sense for checkboxes
+ [x] 4) a numeric field with no limits set doesn’t validate for positive numbers in front after submission, I got an error saying that the value should be less than 0 (sorry I think that was the error I haven’t copied :()
+ [x] 5) I’ve noticed in the DB that the post author is 0 for the default llms_form post types. I think it should be 1.
+ [x] 6) I’ve rapidly submitted a form with checkboxes checked, in the DB, though the checkbox value was empty.
+ [x] 7) When assigning a location to a form, in the “location modal” there are, supposedly, links to the selected product, with the “external link”(open in a new tab) icon, although clicking nothing happens.
+ [x] 8) When you Show keys of the default two options for both radio and checkboxes the "saved values" are empty.
+ [x] 9) Multiple defaults for checkboxes (grouped) are not rendered, only the latest is.

I have to say that I noticed this and noted rapidly because I was actually looking for something else (how the data was saved, retrieved etc) so I might have missed something e.g for the points 4 and 6

### PR DESCRIPTION
**Pull request by eri-trabiccolo**
and use it as post_author for the default form

## Description

Fixes #1577|5
5) I’ve noticed in the DB that the post author is 0 for the default llms_form post types. I think it should be 1.
as we discussed in the comments there below

## How has this been tested?
only manually as of now

## Types of changes
Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [x] My code has been tested.
- [x] My code passes all existing automated tests. 
- [x] My code follows the LifterLMS Coding & Documentation Standards. 

@thomasplevy do you agree with this?

I mean, if you don't have an administrator user or your administrator doesn't have the manage_options cap, the forms author will be 0, we can't search forever for a valid user... right?

### CODE REVIEW

**Comment by thomasplevy** [comment]
@eri-trabiccolo I see no issues with this implementation. Tests and, of course, usage of the function to fix the issue, and I think we're good to go :+1:

**Comment by eri-trabiccolo** [comment]
tests added

on a _different_ subject
I'm not sure why the `create` method allows the update... and why in that case we don't use `wp_update_post`

**Comment by thomasplevy** [comment]
> on a _different_ subject
> I'm not sure why the `create` method allows the update... and why in that case we don't use `wp_update_post`

`wp_update_post()` uses `wp_insert_post()` under the hook, with some additional logic. By using the same function, instead of conditionally using one or the other, we can reduce logic.

The method allows an update so that we can update. I believe I added "update" logic after initially writing the creation method and chose to use the same method in favor of creating a different one because the logic is *nearly* identical. I created the update option so that I could programatically re-install the forms (over top of the existing ones) during development as I rewrote and adjusted the templates.

If the problem is semantic usage of language, I think it's best to think of the `$update` flag as a "recreate" flag, in favor of an "update the form" flag. Notice that you cannot *pass in any updates*, you simply pass in that you *want* to update it. It will then update the existing form post with the template for the requested location.

The most common public use case I can see for this would be developing a "reset" function that could maybe be accessed via an action button on the posts table list or maybe from the admin "tools" ui.

I don't see any issue with using `wp_insert_post()` (when passing an ID) for both creation and update. If you see a specific issue (other than not understanding my reasoning) let me know.

File: tests/phpunit/unit-tests/class-llms-test-forms.php [hunk 1]
```diff
@@ -384,6 +384,95 @@ public function test_create() {
...
+
+		foreach ( $installed as $loc => $id ) {
+			// Expect the first admin to be the forms author.
+			$this->assertEquals( $admins[1], get_post( $id )->post_author, $id );
+		}
+
+		// Clean user* tables.
+		$wpdb->query( "TRUNCATE TABLE $wpdb->users" );
+		$wpdb->query( "TRUNCATE TABLE $wpdb->usermeta" );
+
```
Comment (thomasplevy) [inline tests/phpunit/unit-tests/class-llms-test-forms.php]:
```suggestion
```

We don't need to clean up the test users we created for this test. This won't hurt anything but I'm almost positive it's unnecessary

File: tests/phpunit/unit-tests/class-llms-test-install.php [hunk 2]
```diff
@@ -316,4 +327,57 @@ public function test_install() {
...
+		// Log in as second admin.
+		wp_set_current_user( $admins[1] );
+
+		// Expect the second admin to be returned.
+		$this->assertEquals( $admins[1], LLMS_Install::get_can_install_user_id() );
+
+		// Clean user* tables.
+		$wpdb->query( "TRUNCATE TABLE $wpdb->users" );
+		$wpdb->query( "TRUNCATE TABLE $wpdb->usermeta" );
+
```
Comment (thomasplevy) [inline tests/phpunit/unit-tests/class-llms-test-install.php]:
```suggestion
```

Like above, pretty sure we don't need this cleanup

**Review by thomasplevy — APPROVED** [review]

**Comment by eri-trabiccolo** [comment]
> > on a _different_ subject
> > I'm not sure why the `create` method allows the update... and why in that case we don't use `wp_update_post`
> 
> `wp_update_post()` uses `wp_insert_post()` under the hook, with some additional logic. By using the same function, instead of conditionally using one or the other, we can reduce logic.
> 
> The method allows an update so that we can update. I believe I added "update" logic after initially writing the creation method and chose to use the same method in favor of creating a different one because the logic is _nearly_ identical. I created the update option so that I could programatically re-install the forms (over top of the existing ones) during development as I rewrote and adjusted the templates.
> 
> If the problem is semantic usage of language, I think it's best to think of the `$update` flag as a "recreate" flag, in favor of an "update the form" flag. Notice that you cannot _pass in any updates_, you simply pass in that you _want_ to update it. It will then update the existing form post with the template for the requested location.
> 
> The most common public use case I can see for this would be developing a "reset" function that could maybe be accessed via an action button on the posts table list or maybe from the admin "tools" ui.
> 
> I don't see any issue with using `wp_insert_post()` (when passing an ID) for both creation and update. If you see a specific issue (other than not understanding my reasoning) let me know.

I'm fine with this. Maybe 'recreate' or 'reset' or 'force' would be semantically better? Maybe.
Anyways yeah I know wp_update_post calls wp_insert_post under the hood 😜 + some _minor_ additional logic, I wanted to understand why we decided we didn't need that additional logic. Ok now I get it, and I'm fine with this.

File: tests/phpunit/unit-tests/class-llms-test-forms.php (hunk 1 above)
Comment (eri-trabiccolo) [reply tests/phpunit/unit-tests/class-llms-test-forms.php]: I see what you mean, as of now we wouldn't need that clean up... But I would like to have our tests to not leave anything in the db that could arm another test. If any test would take care of the clean up before quitting (and we generally do that in the `tearDown()`) we wouldn't bore about cleaning at the beginning of a test (when necessary) nor we would spend time on debugging a test that doesn't work as expected (as a result of "environment" changes produced by a previous test).
Am I wrong/missing something obvious/being too much anxious? 😅

**Pull request merged**